### PR TITLE
Compile FTS5 query strings to structured layered queries

### PR DESCRIPTION
**Pull request by serhiy-bzhezytskyy**
The structured Boolean macros execute JSON query trees, but issue #17 originally asked for FTS5's query-string syntax, and it is the only item on that list still open. I offered an implementation there and am following up on it here.

I added a scalar function, fts_parse_query, that compiles an FTS5 query string into the JSON the structured macros already validate and execute. It covers implicit and explicit AND, OR, binary NOT, parentheses, quoted phrases with "" escaping and + concatenation, trailing * prefixes, col : and {col1 col2} : filters, and NEAR(a b, N), whose N maps to near_distance one to one. It is a compiler rather than an engine: there is no new execution path, the index format and the search macros are untouched, and field names are checked by the existing query-time validation.

Where the FTS5 documentation leaves room, I measured SQLite itself and matched it: keywords are recognized in capital letters only, an empty phrase adds no constraint, NEAR distances are re-serialized through an integer, and nesting is bounded at depth 256, the limit FTS5's own error message names. Initial-token anchors (^), column complements (-col :), and phrases inside NEAR are rejected with named errors rather than misparsed.

The tests compile each construct and the precedence cases to exact JSON, assert every rejection message, and check equivalence through the executor against hand-written structured queries. The README documents the function with examples, and the parser benchmarks mirror the analyzer set across bare-term, Boolean, phrase, and NEAR strings.

### CODE REVIEW

**Comment by serhiy-bzhezytskyy** [comment]
@kryonix @lnkuiper — this builds on #56 and follows up on the #17 ask. CI is green, no reviews yet. Tagging in case it's useful.

File: src/fts_query_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+    if (!part.empty()) {
+      words.push_back(part);
+    }
+  }
+  auto leaf = make_uniq<QueryNode>();
+  leaf->query = StringUtil::Join(words, " ");
+  leaf->fields = fields;
+  auto multiple_tokens = quoted || words.size() > 1;
+  if (prefix) {
+    leaf->query_mode = multiple_tokens ? "phrase_prefix" : "autocomplete";
```
Comment (kryonix) [inline src/fts_query_parser.cpp:510]: This lowering changes query results because an unset mode becomes layered `standard` search, which may perform fuzzy, prefix, and substring expansion, while `autocomplete` applies minimum-length, DF-cap, and expansion-limit behavior. Neither contract matches FTS5's exact bare terms and token prefixes.

On a four-document corpus containing `gamma`, `alpha`, `alpine`, and `beta`, SQLite FTS5 returns no rows for `gama` and returns `alpha` plus `alpine` for `a*`. The compiled queries here instead return `gamma` for `gama` and no rows for `a*`.

The compiled representation needs to preserve exact non-prefix and token-prefix semantics—or this function must not claim FTS5 compatibility. Please add execution-level differential tests for an absent near-match term and a one-character prefix; JSON-shape assertions do not catch this boundary failure.

File: src/fts_query_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+    }
+    if (++depth > MAX_QUERY_DEPTH) {
+      error = "fts5 query is nested too deeply";
+      return nullptr;
+    }
+    auto inner = ParsePrimary(narrowed);
+    depth--;
+    return inner;
+  }
+  if (token.type == QueryTokenType::BAREWORD &&
```
Comment (kryonix) [inline src/fts_query_parser.cpp:366]: FTS5 column specifications contain strings, not only barewords, and column names are ASCII case-insensitive. Consequently, `"title" : duck` and quoted names such as `"a b" : alpha` are valid FTS5 syntax. The former currently produces `fts5 query has trailing input`, while `TITLE : duck` compiles to `fields:["TITLE"]` and then fails with `fields contains unknown field: TITLE` for an indexed `title` column. The brace-list path above has the same bareword-only assumption.

Please carry FTS5 column identity through parsing and structured-query validation, including quoted, differently cased, list, and nested-filter variants. SQLite's authoritative grammar models both single names and list members as `STRING`: https://github.com/sqlite/sqlite/blob/master/ext/fts5/fts5parse.y

File: src/fts_query_parser.cpp [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+      quoted = true;
+      parts.push_back(Next().text);
+    } else {
+      error = "fts5 query has a dangling operator";
+      return nullptr;
+    }
+    if (Peek().type == QueryTokenType::STAR) {
+      Next();
+      prefix = true;
+      break;
```
Comment (kryonix) [inline src/fts_query_parser.cpp:491]: Breaking as soon as any `*` is consumed rejects valid FTS5 phrases such as `a* + b*` as trailing input. In SQLite's grammar, each `STRING` in a `+`-concatenated phrase has its own optional prefix marker.

Please either preserve per-component prefix semantics through the compiled representation or explicitly document this as unsupported and return a specific unsupported-syntax error. The README currently advertises quoted-phrase concatenation and trailing prefixes without this limitation.

File: src/fts_query_parser.cpp [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+    if (!input.IsValid()) {
+      error = "fts5 query must not be NULL";
+    } else {
+      error = ParseQuery(input.GetValue().GetString(), query_json);
+    }
+    if (error.empty()) {
+      json_writer.WriteValue(query_json);
+      error_writer.WriteNull();
+    } else {
+      json_writer.WriteNull();
```
Comment (kryonix) [inline src/fts_query_parser.cpp:648]: Returning `NULL` as `query_json` loses the parser error when the function is used as documented:

```sql
search_layered_bm25_query(
    fts_parse_query('(alpha').query_json::JSON
)
```

This reports `query must be a non-null JSON object` instead of `fts5 query contains an unclosed parenthesis`. The same happens through the scalar matching macro and for a NULL input.

Please make the normal composed path propagate the parser's diagnostic. If row-wise non-throwing parsing is also needed, that could remain available through a separately named try-style interface. Cover both table and scalar consumers.

**Review by kryonix — CHANGES_REQUESTED** [review]
Thanks for putting this together. I like the goal, and there is a lot of thoughtful work here around precedence, error handling, nesting limits, tests, and integration with the structured-query macros.

My main hesitation is architectural. Although `fts_parse_query` is exposed as a scalar helper, it effectively introduces a second query-language front end: it owns tokenization, grammar, precedence, an AST, normalization, JSON serialization, diagnostics, and an FTS5 compatibility policy. That is a fairly large new semantic surface in C++, especially when the public structured-query behavior has so far been expressed in pure SQL. Once this becomes public API, its accepted syntax, error messages, and lowering behavior will all be difficult to change without compatibility concerns.

The issues I found also seem related to that boundary rather than isolated parser bugs. The parser can recognize an FTS5 construct correctly but still lower it to a structured-query mode with different semantics. For example, bare terms become layered `standard` searches, which may use fuzzy or substring expansion, while FTS5 terms are exact. Prefixes become `autocomplete`, which has its own length, document-frequency, and expansion limits. This means that even a perfect parser cannot preserve FTS5 behavior until the target JSON/executor contract can represent exact terms and prefixes.

I also think a declarative grammar would make the supported language much easier to review and maintain. The current implementation reconstructs FTS5 context through control flow, which makes distinctions such as a quoted string used as a phrase versus the same string used as a column name easy to miss. SQLite already has a compact, authoritative `fts5parse.y` grammar that captures these rules. I would prefer deriving the supported subset from that grammar—or another checked-in declarative grammar—rather than maintaining an independent handwritten interpretation of FTS5.

DuckDB's PEG implementation initially looked like a natural fit, but after looking into it, I do not think it is currently a drop-in option? @Dtenwolde please and correct me when I'm wrong 👀 Its reusable-looking `PEGParser` parses grammar definitions, while matcher construction remains internal to DuckDB's SQL parser. It also assumes SQL tokenization and case-insensitive literals, which do not directly match FTS5's quoting and uppercase-only operators. Reusing it cleanly would likely require upstream work to expose a stable generic parser API and allow a custom tokenizer. I would not want this extension to depend directly on those parser internals.

A path that would make me more comfortable would be to separate this into two boundaries:

1. First define and test exact FTS5-compatible leaf semantics in the existing SQL executor, including exact terms, token prefixes, phrases, and field identity.
2. Then add the query-string front end using a declarative grammar that lowers into that well-defined representation, with differential tests against SQLite for both parsing and execution.

That does not necessarily mean the final implementation cannot contain C++, or that every FTS5 feature has to be supported immediately. A clearly documented subset with named errors is reasonable. What I would like to avoid is committing to a public FTS5 compiler while both its grammar and its target semantics are still being independently approximated.

So I'm supportive of the direction, but I think this needs an architectural pass before merging. The line comments contain the concrete correctness cases that led me there; resolving the compiler/executor contract first should make the parser implementation and its tests substantially clearer as well.

**Comment by serhiy-bzhezytskyy** [comment]
Thanks — this is very valuable feedback.

Did I read `:510` right — that those `autocomplete` limits cannot be set per leaf (DF cap and expansion limit come from the macro arguments, minimum length is hardcoded), so a parser change cannot reach them?

If you don't object, I'd try your first boundary on its own: per-leaf control of those limits plus the exact and prefix leaf semantics, with the leaf behaviour tested against SQLite — so the API is defined before any front end exists, and exact terms become available to anyone composing `query_json` by hand.

Marking this PR a draft meanwhile.

**Comment by kryonix** [comment]
Yes, that reading is correct. The DF cap and expansion limit currently come from the outer macro arguments and are shared by every leaf. The minimum prefix length is hardcoded in the autocomplete path. There is one further constraint: `term_prefixes` only contains prefixes of lengths 2 and 3, so supporting a one-character FTS5 prefix requires an executor/index-side change as well, not merely another parser or JSON option.

I agree with separating this into a prerequisite PR. I would define `exact` and `prefix` as explicit leaf semantics, though, rather than relying only on per-leaf autocomplete settings to approximate them:

- An exact leaf should return only the analyzed exact term and must not fall back to fuzzy, substring, or prefix expansion when that term is absent.
- A prefix leaf should return every matching token prefix, including one-character prefixes, without silently changing correctness because of autocomplete’s DF or expansion caps.

Per-leaf limits may still be useful as a general structured-query feature, but they should not be what gives an FTS5 leaf its semantics. A finite expansion limit or DF cutoff can still omit rows that SQLite would return. If resource bounds are necessary, I would rather make that an explicit limitation or error than silently call the result FTS5-compatible.

The differential tests should include the existing `gama` and `a*` cases, plus a prefix whose matches exceed `term_limit` and a high-DF prefix. Please test the new modes directly through hand-written `query_json`, independently of the parser.

One related boundary remains for later: FTS5 permits prefixes on individual phrase components, such as `a* + b*`, whereas the current `phrase_prefix` mode only models the final token. It is fine to leave that unsupported initially, but the parser should then report it explicitly.

So yes, no objection—I think establishing those executor semantics first is the right next step.

**Comment by serhiy-bzhezytskyy** [comment]
Example of the four leaf semantics — your four-document corpus (`gamma`, `alpha`, `alpine`, `beta`; `stemmer='none'`, `stopwords='none'` to pair with `unicode61`):

| `query_json` | rows today | SQLite FTS5 | with the new leaves |
|---|---|---|---|
| `{"query":"gama"}` | gamma | none | `exact`: none |
| `{"query":"gama","query_mode":"phrase"}` | gamma | none | unchanged (gamma); see 3 |
| `{"query":"a","query_mode":"autocomplete"}` | none | alpha, alpine | `prefix`: alpha, alpine |
| `{"query":"alpha","fields":["X"]}`, column `x` | error: unknown field X | alpha | alpha |

The test file also covers a prefix with more matches than `term_limit` (40 returned; autocomplete stops at 32) and a high-DF prefix (10 returned; autocomplete drops it), both through hand-written `query_json`.

1. `exact` — the analyzed term only; no fuzzy, substring, or prefix expansion when it is absent. One token per leaf; several tokens are several leaves.
2. `prefix` — every raw token starting with the query, any length including one, weighted as an exact hit; no DF cap, no `term_limit`. One-character prefixes come from a `raw_dict` scan, so the index format does not change.
3. phrases — a one-token phrase stays exact and a one-token phrase-prefix stays a prefix. Today `query_shape` rewrites them to `standard` and `autocomplete`; the README documents that fallback, so this one is your call rather than a fix.
4. field identity — ASCII case-insensitive field names in `fields`, in `field_weights`/`field_b` keys, and in the direct macro's `fields :=`.

If this matches your intent, including the fallback change in 3, I will open the prerequisite PR on these four.

**Comment by kryonix** [comment]
Yes, this matches my intent, and I think these four pieces form a coherent prerequisite PR.

Please include the fallback change in 3. Once the executor has explicit `exact` and `prefix` semantics, a one-token phrase should use `exact`, and a one-token phrase-prefix should use `prefix`. The existing README documents the current fallback, but that fallback is also why `{"query":"gama","query_mode":"phrase"}` still disagrees with FTS5. I would treat this as an intentional compatibility change and call it out clearly.

The one-character `raw_dict` scan sounds reasonable as a correctness fallback that avoids an index-format change. For prefixes of length two or more, I would keep using the prefix sidecar. Please add a plan or performance test for the one-character path, since an uncapped prefix may scan and expand a large part of the dictionary.

The field-identity change also sounds right. Please resolve case-insensitive inputs to the canonical stored field name, including map keys, and reject duplicate case-folded keys such as both `x` and `X` rather than choosing one implicitly.

I agree that `exact` and `prefix` should bypass autocomplete’s DF and expansion limits. Per-leaf limits can remain a separate general feature; they are not needed to define the FTS5-compatible semantics.

With those details, absolutely, go ahead with the prerequisite PR. Thanks again for working on this! 🙏🏻

**Comment by serhiy-bzhezytskyy** [comment]
Opened as #63.
